funding/batch.go

```diff
@@ -0,0 +1,512 @@
+
+	// With the funded PSBT we can now advance the funding state machine of
+	// each of the channels.
+	for _, channel := range b.channels {
+		err = b.cfg.Wallet.PsbtFundingVerify(
+			channel.pendingChanID, unsignedPacket,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("error verifying PSBT: %v", err)
+		}
+	}
+
+	// The funded PSBT was accepted by each of the assemblers, let's now
+	// sign/finalize it.
+	finalizePsbtResp, err := b.cfg.WalletKitServer.FinalizePsbt(
+		ctx, &walletrpc.FinalizePsbtRequest{
+			FundedPsbt: fundPsbtResp.FundedPsbt,
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error finalizing PSBT for batch "+
+			"channel open: %v", err)
+	}
+	finalTx := &wire.MsgTx{}
+	txReader := bytes.NewReader(finalizePsbtResp.RawFinalTx)
+	if err := finalTx.Deserialize(txReader); err != nil {
+		return nil, fmt.Errorf("error parsing signed raw TX: %v", err)
+	}
+	log.Tracef("[batchopenchannel] signed PSBT: %s",
+		base64.StdEncoding.EncodeToString(finalizePsbtResp.SignedPsbt))
+
+	// Advance the funding state machine of each of the channels a last time
+	// to complete the negotiation with the now signed funding TX.
+	for _, channel := range b.channels {
+		err = b.cfg.Wallet.PsbtFundingFinalize(
+			channel.pendingChanID, nil, finalTx,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("error finalizing PSBT: %v", err)
+		}
+	}
+
+	// Now every channel should be ready for the funding transaction to be
+	// broadcast. Let's wait for the updates that actually confirm this
+	// state.
+	b.wg.Add(len(b.channels))
+	b.negotiateErr = nil
+	for _, channel := range b.channels {
+		// Launch another goroutine that waits for the channel pending
+		// response on the update chan.
+		go b.waitForUpdate(channel, false)
+	}
+
+	// Wait for all updates and make sure we're still good to proceed.
+	b.wg.Wait()
+	if b.negotiateErr != nil {
+		return nil, fmt.Errorf("error batch opening channel, final "+
+			"negotiation failed: %v", b.negotiateErr)
+	}
+
+	// Great, we're now finally ready to publish the transaction.
+	err = b.cfg.Wallet.PublishTransaction(finalTx, label)
+	if err != nil {
+		return nil, fmt.Errorf("error publishing final batch "+
+			"transaction: %v", err)
+	}
+	b.didPublish = true
+
+	rpcPoints := make([]*lnrpc.PendingUpdate, len(b.channels))
+	for idx, channel := range b.channels {
+		rpcPoints[idx] = &lnrpc.PendingUpdate{
+			Txid:        channel.chanPoint.Hash.CloneBytes(),
+			OutputIndex: channel.chanPoint.Index,
+		}
+	}
+
+	return rpcPoints, nil
+}
+
+// waitForUpdate waits for an incoming channel update (or error) for a single
+// channel.
+func (b *Batcher) waitForUpdate(channel *batchChannel, firstUpdate bool) {
+	defer b.wg.Done()
+
+	select {
+	// If an error occurs then immediately return the error to the client.
+	case err := <-channel.errChan:
+		log.Errorf("unable to open channel to NodeKey(%x): %v",
+			channel.fundingReq.TargetPubkey.SerializeCompressed(),
+			err)
+		b.negotiateErr = err
```

Comment:
Related to the `errgroup` suggestion above. Though I think the concurrent _replacement_ here by other goroutines is ok, assuming the main one only reads the variable once all the goroutines have excited.